cranelift/zkasm_data/benchmarks/Design.md

```diff
@@ -0,0 +1,112 @@
+# Benchmarking infrastructure design
+
+This documents outlines the design of the infrastructure to be build for ZK WASM benchmarking. The insights we want to gain with benchmarking can be divided into the following two categories:
+
+- **Analysis** targets understanding the performance of zkASM generated by the compiler and tries to answer questions like: Which instructions are executed how often during execution and which are hot instructions? How is performance affected by changes to the compiler? How can the performance of generated zkASM be improved?
+- **Comparison** relates the performance of ZK WASM to the performance of other ZK proving systems and tries to answer questions like: Which proving system is more performant? What are the reasons for differences in performance?
+
+# Analysis
+
+This section describes the technical architecture and user interface of infrastructure and tooling related to *analysis*.
+
+## Instrumentation
+
+Instrumentation is based on logging zkASM instructions by calling Javascript from zkASM. The indirection via Javascript is required as `zkevm-proverjs` executes compiled PIL instead of zkASM. The code for zkASM instruction logging might be inserted to zkASM by enabling a compilation flag or in post processing.
+
+Subsequently raw instruction logs are processed to aggregate data and surface important information. Information to be aggregated includes the number of times an instruction was executed and the number of register allocations. Analyzing the number of register allocations is relevant as [this example](https://github.com/near/wasmtime/issues/181) shows how reducing register allocations can decrease the number of cycles.
+
+## Visualization
+
+Once the data described in the previous section is available and has been used while working on the repository, opportunities for visualizations might be identified. They should help to make the information more easily digestible. Some examples of possible visualizations are:
+
+- The number of times instructions were executed.
+- zkASM instructions and the number of cycles required for their execution.
+- The number of cycles required for benchmarks across different points in the git commit history.
+
+## Relating zkASM instructions to cycles
```

Comment:
What exactly are the instructions that we are going to count? Consider the following [ZKASM line from ZKEVM ROM](https://github.com/0xPolygonHermez/zkevm-rom/blob/877328eff50070426f040a8843b7d6eb3140ef22/main/opcodes/arithmetic.zkasm#L28):
```asm
$               :ADD, MSTORE(SP++), JMP(readCode)
```
This is a valid code that pipelines the execution of 3 different primitive instructions in sequence to use fewer cycles and intermediate registers. I would expect our codegen to use this feature heavily in the future.
How will we track this? Do we canonicalize this into a single compound instruction? Or do we track it as 3 separate instructions?

Reply:
> Do we canonicalize this into a single compound instruction? Or do we track it as 3 separate instructions?

I think both might be interesting, depending on the context of the analysis for which benchmarking is used. Therefore I would suggest to collect data to enable both and provide a UI or CLI switch to choose whether an instruction list is accounted for as a single compound instruction or as separate instructions.